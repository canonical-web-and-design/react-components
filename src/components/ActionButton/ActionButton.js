import classNames from "classnames";
import PropTypes from "prop-types";
import React, { useEffect, useRef, useState } from "react";

export const LOADER_MIN_DURATION = 400; // minimium duration (ms) loader displays
export const SUCCESS_DURATION = 2000; // duration (ms) success tick is displayed

const ActionButton = ({
  appearance = "neutral",
  children,
  className,
  disabled,
  inline,
  loading,
  success,
  ...props
}) => {
  const [height, setHeight] = useState();
  const [width, setWidth] = useState();
  const [showLoader, setShowLoader] = useState(false);
  const [showSuccess, setShowSuccess] = useState(false);
  const ref = useRef(null);

  // Set up loader timer
  useEffect(() => {
    let loaderTimeout;

    if (loading) {
      // Explicitly set button dimensions
      if (ref.current && !!ref.current.getBoundingClientRect()) {
        setHeight(ref.current.getBoundingClientRect().height);
        setWidth(ref.current.getBoundingClientRect().width);
      }
      setShowLoader(true);
    }

    if (!loading && showLoader) {
      loaderTimeout = setTimeout(() => {
        setShowLoader(false);

        if (success) {
          setShowSuccess(true);
        }
      }, LOADER_MIN_DURATION);
    }

    return () => clearTimeout(loaderTimeout);
  }, [loading, showLoader, success]);

  // Set up success timer
  useEffect(() => {
    let successTimeout;

    if (showSuccess) {
      successTimeout = setTimeout(() => {
        setHeight();
        setWidth();
        setShowSuccess(false);
      }, SUCCESS_DURATION);
    }

    return () => clearTimeout(successTimeout);
  }, [showSuccess]);

  const buttonClasses = classNames(className, `p-button--${appearance}`, {
    "is-processing": showLoader || showSuccess,
    "is-disabled": disabled,
    "is-inline": inline,
  });

  const iconClasses = classNames({
    "p-icon--spinner u-animation--spin": showLoader,
    "is-light": appearance === "positive" || appearance === "negative",
    "p-icon--success": showSuccess,
  });

  return (
    <button
      className={buttonClasses}
      disabled={disabled}
      ref={ref}
      style={
        height && width
          ? {
              height: `${height}px`,
              width: `${width}px`,
            }
          : undefined
      }
      {...props}
    >
      {showLoader || showSuccess ? <i className={iconClasses} /> : children}
    </button>
  );
};

ActionButton.propTypes = {
  appearance: PropTypes.string,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  inline: PropTypes.bool,
  loading: PropTypes.bool,
  success: PropTypes.bool,
};

export default ActionButton;
