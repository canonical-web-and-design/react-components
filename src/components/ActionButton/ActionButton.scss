@import "~vanilla-framework/scss/base_typography-max-widths";
@import "~vanilla-framework/scss/base_placeholders";
@import "~vanilla-framework/scss/settings_animations";
@import "~vanilla-framework/scss/settings_colors";
@import "~vanilla-framework/scss/patterns_icons";
@import "~vanilla-framework/scss/base_icon-definitions";

@include vf-b-placeholders;
@include vf-p-icons;

// Identical to Vanilla success icon, but possible to alter tick color
@mixin maas-icon-success($bg-color, $tick-color: $color-x-light) {
  background-image: url("data:image/svg+xml,%3Csvg width='17' height='17' xmlns='http://www.w3.org/2000/svg'%3E%3Cg transform='translate%281 1%29' fill='none' fill-rule='evenodd'%3E%3Ccircle stroke='#{vf-url-friendly-color($bg-color)}' stroke-width='1.5' fill='#{vf-url-friendly-color($bg-color)}' cx='7.25' cy='7.25' r='7.25'/%3E%3Cpath fill='#{vf-url-friendly-color($tick-color)}' d='M11.05 4.173l-.066.058L6.25 8.378l-2.776-2.38-.839.948L6.25 10.75l5.5-5.787-.7-.79z'/%3E%3C/g%3E%3C/svg%3E");
}

.p-icon--success-positive {
  @extend %icon;
  @include maas-icon-success($color-x-light, $color-positive);
}

.p-icon--success-negative {
  @extend %icon;
  @include maas-icon-success($color-x-light, $color-negative);
}

.p-action-button {
  @include vf-animation(
    $property: (
      background-color,
      border,
      opacity
    ),
    $duration: snap
  );

  &.is-loading,
  &.is-success {
    opacity: 1 !important; // Override Vanilla disabled opacity - too much colour changing otherwise
  }
}
