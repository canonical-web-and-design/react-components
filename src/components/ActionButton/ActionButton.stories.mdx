import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import ActionButton from "./ActionButton";

<Meta
  title="ActionButton"
  component={ActionButton}
  argTypes={{
    children: {
      control: {
        type: "text",
      },
    },
  }}
/>

export const Template = (args) => <ActionButton {...args} />;

### ActionButton

This is a not an existing Vanilla component. It can be used to display submitting states for forms or other actions.

### Props

<ArgsTable of={ActionButton} />

### Default

<Canvas>
  <Story
    name="Default"
    args={{ appearance: "positive", children: "Click me!" }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Loading

<Canvas>
  <Story
    name="Loading"
    args={{ appearance: "positive", children: "Click me!", loading: true }}
  >
    {Template.bind({})}
  </Story>
</Canvas>
