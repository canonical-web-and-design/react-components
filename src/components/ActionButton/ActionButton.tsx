import classNames from "classnames";
import PropTypes from "prop-types";
import React, { useEffect, useRef, useState } from "react";
import type { HTMLProps, ReactNode } from "react";

import "./ActionButton.scss";

export const LOADER_MIN_DURATION = 400; // minimium duration (ms) loader displays
export const SUCCESS_DURATION = 2000; // duration (ms) success tick is displayed

type ButtonAppearance = "base" | "neutral" | "positive" | "negative" | "brand";
type Props = {
  children?: ReactNode;
  appearance?: ButtonAppearance;
  className?: string;
  disabled?: boolean;
  inline?: boolean;
  loading?: boolean;
  success?: boolean;
} & HTMLProps<HTMLButtonElement>;

const ActionButton = ({
  appearance = "neutral",
  children,
  className,
  disabled = false,
  inline = false,
  loading = false,
  success = false,
  ...buttonProps
}: Props): JSX.Element => {
  const [height, setHeight] = useState(null);
  const [width, setWidth] = useState(null);
  const [showLoader, setShowLoader] = useState(false);
  const [showSuccess, setShowSuccess] = useState(false);
  const ref = useRef(null);

  // Set up loader timer
  useEffect(() => {
    let loaderTimeout;

    if (loading) {
      // Explicitly set button dimensions
      if (ref.current && !!ref.current.getBoundingClientRect()) {
        setHeight(ref.current.getBoundingClientRect().height);
        setWidth(ref.current.getBoundingClientRect().width);
      }
      setShowLoader(true);
    }

    if (!loading && showLoader) {
      loaderTimeout = setTimeout(() => {
        setShowLoader(false);

        if (success) {
          setShowSuccess(true);
        }
      }, LOADER_MIN_DURATION);
    }

    return () => clearTimeout(loaderTimeout);
  }, [loading, showLoader, success]);

  // Set up success timer
  useEffect(() => {
    let successTimeout;

    if (showSuccess) {
      successTimeout = setTimeout(() => {
        setHeight(null);
        setWidth(null);
        setShowSuccess(false);
      }, SUCCESS_DURATION);
    }

    return () => clearTimeout(successTimeout);
  }, [showSuccess]);

  const buttonClasses = classNames(
    className,
    "p-action-button",
    `p-button--${appearance}`,
    {
      "is-loading": showLoader,
      "is-success": showSuccess,
      "is-disabled": disabled,
      "is-inline": inline,
    }
  );

  const iconClasses = classNames({
    "p-icon--spinner u-animation--spin": showLoader,
    "is-light": appearance === "positive" || appearance === "negative",
    "p-icon--success":
      showSuccess && appearance !== "positive" && appearance !== "negative",
    "p-icon--success-positive": showSuccess && appearance === "positive",
    "p-icon--success-negative": showSuccess && appearance === "negative",
  });

  return (
    <button
      className={buttonClasses}
      disabled={disabled}
      ref={ref}
      style={
        height && width
          ? {
              height: `${height}px`,
              width: `${width}px`,
            }
          : undefined
      }
      {...buttonProps}
    >
      {showLoader || showSuccess ? <i className={iconClasses} /> : children}
    </button>
  );
};

ActionButton.propTypes = {
  appearance: PropTypes.string,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  inline: PropTypes.bool,
  loading: PropTypes.bool,
  success: PropTypes.bool,
};

export default ActionButton;
