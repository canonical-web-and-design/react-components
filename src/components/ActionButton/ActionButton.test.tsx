import { mount } from "enzyme";
import React from "react";
import { act } from "react-dom/test-utils";

import ActionButton, {
  LOADER_MIN_DURATION,
  SUCCESS_DURATION,
} from "./ActionButton";

describe("ActionButton", () => {
  jest.useFakeTimers();

  it("matches loading snapshot", () => {
    const wrapper = mount(<ActionButton loading>Click me</ActionButton>);
    expect(wrapper).toMatchSnapshot();
  });

  it("matches success snapshot", () => {
    const wrapper = mount(<ActionButton loading>Click me</ActionButton>);
    act(() => {
      wrapper.setProps({ loading: false, success: true });
      jest.advanceTimersByTime(LOADER_MIN_DURATION);
      wrapper.update();
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("shows a loader icon for the correct amount of time", () => {
    const wrapper = mount(<ActionButton loading>Click me</ActionButton>);

    act(() => {
      wrapper.setProps({ loading: false, success: true });
      jest.advanceTimersByTime(LOADER_MIN_DURATION - 1);
      wrapper.update();
    });
    expect(wrapper.find(".p-icon--spinner").exists()).toBe(true);

    act(() => {
      jest.advanceTimersByTime(1);
      wrapper.update();
    });
    expect(wrapper.find(".p-icon--spinner").exists()).toBe(false);
  });

  it("shows a loader icon for the correct amount of time", () => {
    const wrapper = mount(<ActionButton loading>Click me</ActionButton>);

    act(() => {
      wrapper.setProps({ loading: false, success: true });
      jest.advanceTimersByTime(LOADER_MIN_DURATION);
      wrapper.update();
    });
    expect(wrapper.find(".p-icon--success").exists()).toBe(true);

    act(() => {
      jest.advanceTimersByTime(SUCCESS_DURATION);
      wrapper.update();
    });
    expect(wrapper.find(".p-icon--success").exists()).toBe(false);
  });

  it("shows the positive success icon if appearance is positive", () => {
    const wrapper = mount(
      <ActionButton appearance="positive" loading>
        Click me
      </ActionButton>
    );

    act(() => {
      wrapper.setProps({ loading: false, success: true });
      jest.advanceTimersByTime(LOADER_MIN_DURATION);
      wrapper.update();
    });
    expect(wrapper.find(".p-icon--success-positive").exists()).toBe(true);
    expect(wrapper.find(".p-icon--success-negative").exists()).toBe(false);
    expect(wrapper.find(".p-icon--success").exists()).toBe(false);
  });

  it("shows the negative success icon if appearance is negative", () => {
    const wrapper = mount(
      <ActionButton appearance="negative" loading>
        Click me
      </ActionButton>
    );

    act(() => {
      wrapper.setProps({ loading: false, success: true });
      jest.advanceTimersByTime(LOADER_MIN_DURATION);
      wrapper.update();
    });
    expect(wrapper.find(".p-icon--success-negative").exists()).toBe(true);
    expect(wrapper.find(".p-icon--success-positive").exists()).toBe(false);
    expect(wrapper.find(".p-icon--success").exists()).toBe(false);
  });
});
