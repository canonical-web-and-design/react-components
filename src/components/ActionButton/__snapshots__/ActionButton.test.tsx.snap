// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton matches loading snapshot 1`] = `
<ActionButton
  loading={true}
>
  <button
    className="p-action-button p-button--neutral is-loading"
    disabled={false}
  >
    <i
      className="p-icon--spinner u-animation--spin"
    />
  </button>
</ActionButton>
`;

exports[`ActionButton matches success snapshot 1`] = `
<ActionButton
  loading={false}
  success={true}
>
  <button
    className="p-action-button p-button--neutral is-success"
    disabled={false}
  >
    <i
      className="p-icon--success"
    />
  </button>
</ActionButton>
`;
